config/database.yml

```diff
@@ -11,6 +11,10 @@ default: &default
   connect_timeout: 5
   variables:
     statement_timeout: 5000 # ms
+  sslmode: <%= ENV.fetch('DATABASE_SSL_MODE', ENV.fetch('DATABASE_SSL_CA', nil) ? 'verify-full' : 'disable') %>
+  sslrootcert: <%= ENV['DATABASE_SSL_CA'] %>
+  sslcert: <%= ENV['DATABASE_SSL_CERT'] %>
+  sslkey: <%= ENV['DATABASE_SSL_KEY'] %>
```

Comment:
I have a concern here. Previously these options were not being set. For example in the past we  would not have `sslrootcert` at all. Not in case secret does not exist, that option will be equivalent to `sslrootcert: null`.

This might be fine in case db client ignores the option when it is `null`. But if it instead considers that there is no root CA this would be a change in behavior where previously the operating system root CA was used (I assume).

So existing setups where OS CA was used for ssl connection would start failing for no CA being configured.

I hope my explanation makes sense.

Reply:
I don't think there's a breaking change because SSL for external DB's wasn't officially supported before, so no client should be using it. And even if somebody tried to connect to a TLS protected DB, the operator doesn't support it at all yet, so I don't know how would they have done it.

Reply:
They could have had a certificate already in common CA. Like from LetsEncrypt. Dunno if it supports databases. But the thing is that they don't necessarily need a custom CA to use ssl. And we shouldn't enforce that.

But it depends on how the client actually behaves.

Reply:
No, but if they don't provide a CA or a client crt or key, how would the client know it has to use SSL at all? the answer is, in that scenario, they would need to set `sslmode` to `required`, So they need the operator, thus the env vars.

Reply:
They can put it in DATABASE_URL which was already available in the past.

Reply:
Whether the operator should use env vars or add arguments to the URI I think it's another debate. I would say, whatever operator engineers think it's better or easier for them. Do you suggest to close these PRs?

Reply:
My point was that this could already be used. And also people may want to use SSL without specifying a certificate in case DB uses OS approved CAs. For example Amazon RDS, maybe it works with default CAs. Why should one add a secret for them?

I'm suggesting to make sure that we don't brake this use case. What you said that these variables are useful for the operator makes sense.

I also hate such PRs because they take too much effort for something technically simple.

Reply:
For that use case, they can still append `?ssl_mode=required` to `DATABASE_URL` and it should work

Reply:
but will `sslrootcert` keep using the default? I want to test this, perhaps later today with [s_client](https://dba.stackexchange.com/a/243139/80702), adding the cert to RHEL CAs and running the gem with the option set to `nil` and option not set... to see for any differences.

But feel free to do it before me if you have time for it :)

Reply:
I started postgres with SSL. First create certificates according to https://www.postgresql.org/docs/current/ssl-tcp.html

Then start postgres container:
```
podman run "-d"                     "-p"                     "5432:5432"                     "-e"                     "POSTGRES_USER=postgres"                     "-e"                     "POSTGRES_DB=zync"                     "--name"                     "postgres14-zync"                     "cimg/postgres:14.12" -c ssl=on
```

This fails to start for no cert files. Cpy the cert files:
```
podman cp ~/ca/postgres.crt postgres14-zync:/var/lib/postgresql/data/server.crt
podman cp ~/ca/postgres.key postgres14-zync:/var/lib/postgresql/data/server.key
```

Now start postgres
```
podman start postgres14-zync
```

Now try to connect:
```sh
DATABASE_URL="postgresql://postgres:@127.0.0.1/3scale_system_development?sslmode=verify-full" rails db:setup
```

It would fail with something like:
```
connection to server at "127.0.0.1", port 5432 failed: could not initiate GSSAPI security context: Unspecified GSS failure.  Minor code may provide more information: Server postgres/localhost@IPA.REDHAT.COM not found in Kerberos database (PG::ConnectionBad)
connection to server at "127.0.0.1", port 5432 failed: root certificate file "/home/avalon/.postgresql/root.crt" does not exist
Either provide the file, use the system's trusted roots with sslrootcert=system, or change sslmode to disable server certificate verification.
```

What works without modifying the image (in case DB uses a certificate signed by a trusted authority)
```
DATABASE_URL="postgresql://postgres:@localhost/3scale_system_development?sslmode=verify-full&sslrootcert=system" rails db:setup
```

while on it, to add a cert to fedora trusted list, put the certificate inside `/etc/pki/ca-trust/source/anchors/` and then run `sudo update-ca-trust`.

But it all seems unlikely that anybody used that. Because Amazon RDS seems to use its own CA. And it is unlikely anybody purchases certificates for an internal DB from public authorities.

Still how about having this whole block of 4 options be added to config only if `DATABASE_SSL_CA` wa set?

 ```suggestion
  <% if ENV.fetch('DATABASE_SSL_CA', false) %>
  sslmode: <%= ENV.fetch('DATABASE_SSL_MODE', 'verify-full') %>
  sslrootcert: <%= ENV['DATABASE_SSL_CA'] %>
  sslcert: <%= ENV['DATABASE_SSL_CERT'] %>
  sslkey: <%= ENV['DATABASE_SSL_KEY'] %>
  <% end %>
```

Reply:
I've been doing some tests:

1. In all scenarios I tested, there's no difference between not providing options like `sslrootcert` or providing them `nil`
2. If I install my self-signed certificate with `update-ca-trust`. This works fine:
```
DATABASE_URL="postgresql://postgres:@localhost/3scale_system_development?sslmode=verify-full&sslrootcert=system" bundle exec rails boot
```

> Still how about having this whole block of 4 options be added to config only if `DATABASE_SSL_CA` wa set?

I don't like the idea. First, because according to my tests it makes no difference; and second, because there might be some scenario where any other var like `DATABASE_SSL_MODE` is set but `DATABASE_SSL_CA` isn't.

Reply:
Oh, good, I didn't test this. In this case everything is fine. Thank you for bearing with me!